File: leetcode-string-processing-solutions-pro/test_string_manipulation.cpp
#include "string_manipulation.cpp"  // Include your implementation
#include <cassert>
#include <iostream>

void testReverseString() {
  assert(reverseString("hello") == "olleh");
  assert(reverseString("") == "");
  assert(reverseString("a") == "a");
}

void testIsPalindrome() {
  assert(isPalindrome("racecar") == true);
  assert(isPalindrome("hello") == false);
  assert(isPalindrome("") == true);
  assert(isPalindrome("a") == true);
}

void testLongestPalindromeSubstring(){
    assert(longestPalindromeSubstring("babad") == "bab" || longestPalindromeSubstring("babad") == "aba");
    assert(longestPalindromeSubstring("cbbd") == "bb");
    assert(longestPalindromeSubstring("a") == "a");
    assert(longestPalindromeSubstring("") == "");

}

void testArePermutations() {
  assert(arePermutations("abc", "bca") == true);
  assert(arePermutations("abc", "abcd") == false);
  assert(arePermutations("", "") == true);
  assert(arePermutations("a", "a") == true);
}

void testRemoveDuplicateChars() {
  assert(removeDuplicateChars("abbcdeff") == "abcdef");
  assert(removeDuplicateChars("") == "");
  assert(removeDuplicateChars("a") == "a");
  assert(removeDuplicateChars("aabbccddeeff") == "abcdef");

}


int main() {
  testReverseString();
  testIsPalindrome();
  testLongestPalindromeSubstring();
  testArePermutations();
  testRemoveDuplicateChars();
  std::cout << "All test cases passed!" << std::endl;
  return 0;
}

File: algorithm-string-processing-implementation-pro/test_string_manipulation.py
```python
import unittest
from string_manipulation import reverse_string, is_palindrome, longest_palindrome_substring, string_compression, are_anagrams

class TestStringManipulation(unittest.TestCase):
    def test_reverse_string(self):
        self.assertEqual(reverse_string("hello"), "olleh")
        self.assertEqual(reverse_string(""), "")
        self.assertEqual(reverse_string("a"), "a")

    def test_is_palindrome(self):
        self.assertTrue(is_palindrome("racecar"))
        self.assertTrue(is_palindrome("A man, a plan, a canal: Panama"))
        self.assertFalse(is_palindrome("hello"))

    def test_longest_palindrome_substring(self):
        self.assertEqual(longest_palindrome_substring("babad"), "bab" or "aba") #Multiple possibilities
        self.assertEqual(longest_palindrome_substring("cbbd"), "bb")
        self.assertEqual(longest_palindrome_substring("a"), "a")

    def test_string_compression(self):
        self.assertEqual(string_compression("aabcccccaaa"), "a2b1c5a3")
        self.assertEqual(string_compression("abcd"), "abcd")
        self.assertEqual(string_compression(""), "")

    def test_are_anagrams(self):
        self.assertTrue(are_anagrams("listen", "silent"))
        self.assertTrue(are_anagrams("abc", "bca"))
        self.assertFalse(are_anagrams("abc", "abd"))

if __name__ == '__main__':
    unittest.main()
```